boa/vm/py_evm.py

```diff
@@ -0,0 +1,549 @@
+
+        if bytecode in cls.env._code_registry:
+            target = cls.env._code_registry[bytecode].deployer.at(contract_address)
+            target.created_from = Address(msg.sender)
+            cls.env.register_contract(contract_address, target)
+
+        return computation
+
+    @classmethod
+    def apply_computation(cls, state, msg, tx_ctx, **kwargs):
+        addr = msg.code_address
+        contract = cls.env._lookup_contract_fast(addr) if addr else None
+
+        def finalize(c):
+            if c.is_error:
+                # After the computation is applied with an error the state is
+                # reverted. Before the revert, save the contract repr for the
+                # error message
+                c._contract_repr_before_revert = repr(contract)
+            return c
+
+        if contract is None or not cls.env.evm._fast_mode_enabled:
+            # print("SLOW MODE")
+            computation = super().apply_computation(state, msg, tx_ctx, **kwargs)
+            return finalize(computation)
+
+        with cls(state, msg, tx_ctx) as computation:
+            try:
+                if getattr(msg, "_ir_executor", None) is not None:
+                    # print("MSG HAS IR EXECUTOR")
+                    # this happens when bytecode is overridden, e.g.
+                    # for injected functions. note ir_executor is (correctly)
+                    # used for the outer computation only because on subcalls
+                    # a clean message is constructed for the child computation
+                    msg._ir_executor.exec(computation)
+                else:
+                    # print("REGULAR FAST MODE")
+                    contract.ir_executor.exec(computation)
+            except Halt:
+                pass
+
+        # return computation outside of with block; computation.__exit__
+        # swallows exceptions (including Revert).
+        return finalize(computation)
+
+
+# Message object with extra attrs we can use to thread things through
+# the execution context.
+class FakeMessage(Message):
+    def __init__(
+        self,
+        *args,
+        ir_executor=None,
+        fake_codesize=None,
+        start_pc=0,
+        contract=None,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self._ir_executor = ir_executor
+        self._fake_codesize = fake_codesize
+        self._start_pc = start_pc
+        self._contract = contract
+
+
+class PyEVM:
+    def __init__(self, env, fast_mode_enabled=False, fork_try_prefetch_state=False):
+        self.chain = _make_chain()
+        self.env = env
+        self._fast_mode_enabled = fast_mode_enabled
+        self._fork_try_prefetch_state = fork_try_prefetch_state
+        self._init_vm()
+
+    def _init_vm(self, account_db_class=AccountDB):
+        self.vm = self.chain.get_vm()
+        self.vm.__class__._state_class.account_db_class = account_db_class
+
+        self.vm.patch = VMPatcher(self.vm)
```

Comment:
we can probably promote VMPatcher to `self.patch` instead of inside of `self.vm`